### ISSUE TITLE
Warnings in IAR compiler

### ISSUE THREAD

**Issue by richnwhit** [bblanchon/ArduinoJson#1233]
I know the IAR Compiler isn't a supported option, but the project is working fine on it and compiles with just two warnings, so I thought it worth reporting these especially as they look sensible.

```c++
template <typename TAdaptedString>  
inline bool slotSetKey(VariantSlot* var, TAdaptedString key, MemoryPool* pool,
                       storage_policy::decide_at_runtime) {
  if (key.isStatic()) {
    return slotSetKey(var, key, pool, storage_policy::store_by_address());
  } else {
    return slotSetKey(var, key, pool, storage_policy::store_by_copy());
  }
  return true;       // <-- warning here
}
````

Warning : Statement is unreachable

```c++
void visitPositiveInteger(UInt value) {
    if (value <= 0x7F) {
      writeInteger(uint8_t(value));
    } else if (value <= 0xFF) {
      writeByte(0xCC);
      writeInteger(uint8_t(value));
    } else if (value <= 0xFFFF) {
      writeByte(0xCD);
      writeInteger(uint16_t(value));
    } else if (value <= 0xFFFFFFFF) {     // <-- Warning here
      writeByte(0xCE);
      writeInteger(uint32_t(value));
    }
#if ARDUINOJSON_USE_LONG_LONG
    else {
      writeByte(0xCF);
      writeInteger(uint64_t(value));
    }
#endif
  }
```

Warning: Pointless integer comparison, the result is always true

**Comment by bblanchon** [comment]
Hi @richnwhit,

Thank you for reporting these warnings.
IAR compiler **is** a supported option.
Can you give me the exact version/architecture of the compiler so I can try to repro here?

Best regards,
Benoit

**Comment by richnwhit** [comment]
Benoit,

Sorry I misspoke - I am using Embedded Workbench, which is indeed supported. But I am using the ARM compiler (version 8.40.1), which you may not be supporting (it isn't free).

The warnings looked pretty sensible and straightforward (I'm surprised that other tools aren't complaining); if you want me to check an unofficial version for you on the IAR compiler, please let me know.

BTW I did successfully implement a memory test, by storing a dummy value and then checking that it is present.

Thanks a lot for your excellent work,

Richard

**Comment by bblanchon** [comment]
@richnwhit, can you check out the branch "issue1233" to verify that the warnings are gone?

**Comment by richnwhit** [comment]
Benoit,

Would it be possible for you to create the single-header version for me? I can try to build the library from source in ARM compiler, but that's not what I was doing, and I may get different issues.

Thanks,

Richard

**Comment by bblanchon** [comment]
[ArduinoJson-v6.15.0-6-g40d1cfe7.zip](https://github.com/bblanchon/ArduinoJson/files/4442857/ArduinoJson-v6.15.0-6-g40d1cfe7.zip)

**Comment by richnwhit** [comment]
Benoît,

That's fixed it, thank you!

Richard

**Comment by bblanchon** [comment]
Shipped in [ArduinoJson 6.15.1](https://github.com/bblanchon/ArduinoJson/releases/tag/v6.15.1)
